cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicaProvider.java

```diff
@@ -0,0 +1,165 @@
+
+        // Retrieve map of primary token ranges to endpoints that describe the ring topology
+        Map<List<String>, List<String>> rangeToEndpointMappings = storage.getRangeToEndpointWithPortMap(keyspace);
+        // Pending ranges include bootstrap tokens and leaving endpoints as represented in the Cassandra TokenMetadata
+        Map<List<String>, List<String>> pendingRangeMappings = storage.getPendingRangeToEndpointWithPortMap(keyspace);
+
+        Stream<String> hostsStream = Stream.concat(rangeToEndpointMappings.values().stream().flatMap(List::stream),
+                                                   pendingRangeMappings.values().stream().flatMap(List::stream));
+
+        Map<String, String> hostToDatacenter = groupHostsByDatacenter(hostsStream);
+
+        // Retrieve map of all token ranges (pending & primary) to endpoints
+        List<TokenRangeReplicasResponse.ReplicaInfo> writeReplicas =
+        writeReplicasFromPendingRanges(rangeToEndpointMappings,
+                                       pendingRangeMappings,
+                                       hostToDatacenter,
+                                       partitioner,
+                                       keyspace);
+
+        return new TokenRangeReplicasResponse(
+                writeReplicas,
+                mappingsToUnwrappedReplicaSet(rangeToEndpointMappings, hostToDatacenter, partitioner));
+    }
+
+    private List<TokenRangeReplicasResponse.ReplicaInfo>
+    writeReplicasFromPendingRanges(Map<List<String>, List<String>> readReplicaMappings,
+                                   Map<List<String>, List<String>> pendingRangeMappings,
+                                   Map<String, String> hostToDatacenter,
+                                   Partitioner partitioner,
+                                   String keyspace)
+    {
+        LOGGER.debug("Pending token ranges for keyspace={}, pendingRangeMappings={}", keyspace, pendingRangeMappings);
+        // Merge primary and pending replicas to generate candidates for write-replicas
+        List<TokenRangeReplicas> replicas = Stream.concat(
+                                                  readReplicaMappings.entrySet().stream(),
+                                                  pendingRangeMappings.entrySet().stream())
+                                                  .map(entry ->
+                                                       new TokenRangeReplicas(
+                                                               new BigInteger(entry.getKey().get(0)),
+                                                               new BigInteger(entry.getKey().get(1)),
+                                                               partitioner,
+                                                               new HashSet<>(entry.getValue())))
+                                                  .collect(Collectors.toList());
+
+        // Candidate write-replica mappings are normalized by removing overlapping ranges
```

Comment:
It took me some time to finally realize what the `normalize` and `deoverlap` methods do. I think adding the following comments should help future readers. 

```suggestion
        // Candidate write-replica mappings (merged from natural and pending ranges) are normalized by consolidate overlapping ranges
        // For an overlapping range that is included in both natural and pending ranges, say R_natural and R_pending 
        // (where R_natural == R_pending), the replicas of both R_natural and R_pending should receive writes. 
        // Therefore, the write-replicas of such range is the union of both replica sets. 
        // The procedure (esp. normalize method) implements the consolidation process.
```